### ISSUE TITLE
ext/File-Find/t/taint.t: intermittent failure, possibly due to stale file handle

### ISSUE THREAD

**Issue by jkeenan** [Perl/perl5#20734]
Recent work done on the tests in `ext/File-Find/` may have introduced intermittent test failures in `ext/File-Find/t/*.t`.  These test failures are appearing in test environments which are, I suspect, resource-challenged.  One of them appears to be a Raspberry Pi running Linux; the other is my OpenBSD-6.9 VM.

**Linux**

https://perl5.test-smoke.org/report/5029413

This is one of George Greer's smokers, which was testing one of @khwilliamson's branches (now superseded).

```
Automated smoke report for branch smoke-me/khw-env 5.37.8 patch d26d6c57c7f3c558bf9b5e29bab7752f9067c18f 
v5.37.8-152-gd26d6c57c7.
develpi.local: ARMv7 Processor rev 3 (v7l) (300 MHz) (armv7l/4 cpus)
    on        linux 5.10.103-v7l+ [Raspbian GNU/Linux 10.13]
    using     ccache gcc version 8.3.0 (*1)
...

Test failures:
  ../ext/File-Find/t/taint.t (Non-zero exit status: 116 Bad plan.  You planned 45 tests but ran 1.)
      [stdio] -Dcc="ccache gcc"
```

In the [log file](https://m-l.org/~perl/smoke/perl/arm-linux-gnueabihf-thread-multi-64int/smoke-me_gcc_quick/Karl%20Williamson/logd26d6c57c7f3c558bf9b5e29bab7752f9067c18f.log.gz) which George provides we see this:
```
Can't cd to .. from ./FF_find_t_9kHmlb/for_find/fb/fbc: Stale file handle at /home/perl/smoke/perl/s
moke-me_gcc_quick/build/ext/File-Find/../../lib/File/Find.pm line 477.
# Looks like your test exited with 116 just after 1.
# locales available: C C.UTF-8 POSIX en_US.utf8
[2023-01-20 18:20:01-0500] [died test] ../ext/File-Find/t/taint.t..................................
[2023-01-20 18:20:01-0500] [TAP-error test] ../ext/File-Find/t/taint.t..............................
....
make: *** [makefile:865: test_harness] Error 1

error while running harness target 'test_harness': 2 at /home/perl/smoke/ts/Test-Smoke-1.72-custom/b
in/../lib/Test/Smoke.pm line 195.
[2023-01-20 18:20:02-0500] 
[2023-01-20 18:20:02-0500] ../ext/File-Find/t/taint.t..................................FAILED
[2023-01-20 18:20:02-0500]     Non-zero exit status: 116
[2023-01-20 18:20:02-0500] ../ext/File-Find/t/taint.t..................................FAILED
[2023-01-20 18:20:02-0500]     Bad plan.  You planned 45 tests but ran 1.
```
Note that mention of a **Stale file handle**

**OpenBSD-6.9**

I've observed failures in `taint.t` twice in recent weeks in this VM:  once while doing a regular smoke-test run, once while testing the perl-5.37.8 tarball earlier this week.  The reports are similar, so let's present the smoke-test run.

https://perl5.test-smoke.org/report/5029144

```
Automated smoke report for branch blead 5.37.8 patch bc01770b32f9d9bf9d313d2eec370c41822cd61c 
v5.37.7-146-gbc01770b32.
perl-reporter-11: Intel(R) Xeon(R) CPU E31220 @ 3.10GHz (3093 MHz) (amd64/2 cpus)
    on        openbsd 6.9
    using     cc version OpenBSD Clang 10.0.1 (*1)
...
Test failures:
  ../ext/File-Find/t/taint.t (Non-zero exit status: 2 Bad plan.  You planned 45 tests but ran 0.)
      [stdio] -Duseithreads -Accflags="-DNO_LOCALE_NUMERIC -DNO_LOCALE_COLLATE -fno-delete-null-pointer-checks"
```
From the log file:
```
Can't cd to .. from ./FF_find_t_uHzKN6/for_find/fc/fca: No such file or directory at /home/jkeenan/p
5smoke/perl-current/ext/File-Find/../../lib/File/Find.pm line 477.
# Looks like your test exited with 2 before it could output anything.
[2023-01-16 01:34:06+0000] [died test] ../ext/File-Find/t/taint.t..................................
[2023-01-16 01:34:06+0000] [TAP-error test] ../ext/File-Find/t/taint.t..............................
....
*** Error 1 in /home/jkeenan/p5smoke/perl-current (makefile:851 'test_harness')

error while running harness target 'test_harness': 2 at /home/jkeenan/p5smoke/lib/Test/Smoke/App/RunSmoke.pm line 118.
[2023-01-16 01:34:07+0000] 
[2023-01-16 01:34:07+0000] ../ext/File-Find/t/taint.t..................................FAILED
[2023-01-16 01:34:07+0000]     Non-zero exit status: 2
[2023-01-16 01:34:07+0000] ../ext/File-Find/t/taint.t..................................FAILED
[2023-01-16 01:34:07+0000]     Bad plan.  You planned 45 tests but ran 0.
```
This is similar to George's report except that it lacks the system warning about "Stale file handle".

This is the first time I've ever heard of a stale file handle.  DDG search turned up some interesting articles, but this is beyond my expertise.

@demerphq, the revisions you did in `ext/File-Find` earlier this month certainly look plausible, but could you take a look to see if they might have generated these failures?

Thank you very much.

**Comment by demerphq** [comment]
On Mon, 23 Jan 2023, 04:21 James E Keenan, ***@***.***> wrote:

> Recent work done on the tests in ext/File-Find/ may have introduced
> intermittent test failures in ext/File-Find/t/*.t. These test failures
> are appearing in test environments which are, I suspect,
> resource-challenged.
>
Are the tests being run in parallel or in series?

> One of them appears to be a Raspberry Pi running Linux; the other is my
> OpenBSD-6.9 VM.
>
> *Linux*
>
> https://perl5.test-smoke.org/report/5029413
>
> This is one of George Greer's smokers, which was testing one of
> @khwilliamson <https://github.com/khwilliamson>'s branches (now
> superseded).
>
> Automated smoke report for branch smoke-me/khw-env 5.37.8 patch d26d6c57c7f3c558bf9b5e29bab7752f9067c18f
> v5.37.8-152-gd26d6c57c7.
> develpi.local: ARMv7 Processor rev 3 (v7l) (300 MHz) (armv7l/4 cpus)
>     on        linux 5.10.103-v7l+ [Raspbian GNU/Linux 10.13]
>     using     ccache gcc version 8.3.0 (*1)
> ...
>
> Test failures:
>   ../ext/File-Find/t/taint.t (Non-zero exit status: 116 Bad plan.  You planned 45 tests but ran 1.)
>       [stdio] -Dcc="ccache gcc"
>
> In the log file
> <https://m-l.org/~perl/smoke/perl/arm-linux-gnueabihf-thread-multi-64int/smoke-me_gcc_quick/Karl%20Williamson/logd26d6c57c7f3c558bf9b5e29bab7752f9067c18f.log.gz>
> which George provides we see this:
>
> Can't cd to .. from ./FF_find_t_9kHmlb/for_find/fb/fbc: Stale file handle at /home/perl/smoke/perl/s
> moke-me_gcc_quick/build/ext/File-Find/../../lib/File/Find.pm line 477.
> # Looks like your test exited with 116 just after 1.
> # locales available: C C.UTF-8 POSIX en_US.utf8
> [2023-01-20 18:20:01-0500] [died test] ../ext/File-Find/t/taint.t..................................
> [2023-01-20 18:20:01-0500] [TAP-error test] ../ext/File-Find/t/taint.t..............................
> ....
> make: *** [makefile:865: test_harness] Error 1
>
> error while running harness target 'test_harness': 2 at /home/perl/smoke/ts/Test-Smoke-1.72-custom/b
> in/../lib/Test/Smoke.pm line 195.
> [2023-01-20 18:20:02-0500]
> [2023-01-20 18:20:02-0500] ../ext/File-Find/t/taint.t..................................FAILED
> [2023-01-20 18:20:02-0500]     Non-zero exit status: 116
> [2023-01-20 18:20:02-0500] ../ext/File-Find/t/taint.t..................................FAILED
> [2023-01-20 18:20:02-0500]     Bad plan.  You planned 45 tests but ran 1.
>
> Note that mention of a *Stale file handle*
>
> *OpenBSD-6.9*
>
> I've observed failures in taint.t twice in recent weeks in this VM: once
> while doing a regular smoke-test run, once while testing the perl-5.37.8
> tarball earlier this week. The reports are similar, so let's present the
> smoke-test run.
>
> https://perl5.test-smoke.org/report/5029144
>
> Automated smoke report for branch blead 5.37.8 patch bc01770b32f9d9bf9d313d2eec370c41822cd61c
> v5.37.7-146-gbc01770b32.
> perl-reporter-11: Intel(R) Xeon(R) CPU E31220 @ 3.10GHz (3093 MHz) (amd64/2 cpus)
>     on        openbsd 6.9
>     using     cc version OpenBSD Clang 10.0.1 (*1)
> ...
> Test failures:
>   ../ext/File-Find/t/taint.t (Non-zero exit status: 2 Bad plan.  You planned 45 tests but ran 0.)
>       [stdio] -Duseithreads -Accflags="-DNO_LOCALE_NUMERIC -DNO_LOCALE_COLLATE -fno-delete-null-pointer-checks"
>
> From the log file:
>
> Can't cd to .. from ./FF_find_t_uHzKN6/for_find/fc/fca: No such file or directory at /home/jkeenan/p
> 5smoke/perl-current/ext/File-Find/../../lib/File/Find.pm line 477.
> # Looks like your test exited with 2 before it could output anything.
> [2023-01-16 01:34:06+0000] [died test] ../ext/File-Find/t/taint.t..................................
> [2023-01-16 01:34:06+0000] [TAP-error test] ../ext/File-Find/t/taint.t..............................
> ....
> *** Error 1 in /home/jkeenan/p5smoke/perl-current (makefile:851 'test_harness')
>
> error while running harness target 'test_harness': 2 at /home/jkeenan/p5smoke/lib/Test/Smoke/App/RunSmoke.pm line 118.
> [2023-01-16 01:34:07+0000]
> [2023-01-16 01:34:07+0000] ../ext/File-Find/t/taint.t..................................FAILED
> [2023-01-16 01:34:07+0000]     Non-zero exit status: 2
> [2023-01-16 01:34:07+0000] ../ext/File-Find/t/taint.t..................................FAILED
> [2023-01-16 01:34:07+0000]     Bad plan.  You planned 45 tests but ran 0.
>
> Why does it say the test failed twice?  Did it actually launch the test
twice?

This is similar to George's report except that it lacks the system warning
> about "Stale file handle".
>
> This is the first time I've ever heard of a stale file handle. DDG search
> turned up some interesting articles, but this is beyond my expertise.
>
> @demerphq <https://github.com/demerphq>, the revisions you did in
> ext/File-Find earlier this month certainly look plausible, but could you
> take a look to see if they might have generated these failures
>

I'll take a look, but off the top of my head I can't see how this patch
would trigger an error like this.

The duplicate fail report in that test suggests something else is at play.
Was this test run under 'test_harness' or 'test'?

Do smokers set anything that might derandomize the tests?

Could the smoke framework contain a big and run the tests twice?

If the answer to those last two questions is yes then these failures would
be quite understandable.

I think we need to make harness report how many jobs it was running when it
finishes.

Yves

>

**Comment by demerphq** [comment]
@jkeenan were you able to answer any of my questions? I haven't poked into how Test::Smoke works yet.

**Comment by jkeenan** [comment]
> Are the tests being run in parallel or in series?

Parallel.  My smoke-testing rigs are all set up to run via `make test_harness` only; I don't do them via `make test`.  And the instance where I was testing the tarball was a direct invocation of `make test_harness`.

...

> Why does it say the test failed twice? Did it actually launch the test
> twice? 

No, that's just a Test-Smoke peculiarity.

...

> I'll take a look, but off the top of my head I can't see how this patch would trigger an error like this. The duplicate fail report in that test suggests something else is at play. Was this test run under 'test_harness' or 'test'?

See above.

> Do smokers set anything that might derandomize the tests? 

No.

> Could the smoke framework contain a bug and run the tests twice?

AFAIK, no.  In our more underpowered machines used for smoking, it is not uncommon to see test failures due to a timeout or to a temporarly loss of network connectivity.  In such cases, we have never suspected that the failed file was being run twice.  For example, here is a section of a Test-Smoke log from the same machine but a different run.
```
[2022-11-11 15:17:54+0000] [died test] ../dist/Time-HiRes/t/alarm.t................................
*** Error 1 in /home/jkeenan/p5smoke/perl-current (makefile:842 'test_harness')

error while running harness target 'test_harness': 2 at /home/jkeenan/p5smoke/lib/Test/Smoke/App/RunSmoke.pm line 118.
[2022-11-11 15:17:55+0000]
[2022-11-11 15:17:55+0000] ../dist/Time-HiRes/t/alarm.t................................FAILED
[2022-11-11 15:17:55+0000]     2
[2022-11-11 15:17:55+0000] ../dist/Time-HiRes/t/alarm.t................................FAILED
[2022-11-11 15:17:55+0000]     Non-zero exit status: 1
```
> If the answer to those last two questions is yes then these failures would be quite understandable. I think we need to make harness report how many jobs it was running when it finishes. Yves
> […](#)

**Comment by jkeenan** [comment]
Another instance of this failure on OpenBSD-6.9 VM:
https://perl5.test-smoke.org/report/5029819

**Comment by demerphq** [comment]
@jkeenan I believe that #20753 fixes this.

**Comment by khwilliamson** [comment]
FWIW, I had to make `t/29a_upcopy.t` execute  entirely standalone to get it to reliably pass.  Making it not executable with other files in the directory was not good enough

**Comment by demerphq** [comment]
> FWIW, I had to make t/29a_upcopy.t execute entirely standalone to get it to reliably pass. Making it not executable with other files in the directory was not good enough

@khwilliamson  can you file a ticket for that preferably with test output? This thread is about File-Find, that test is in Tie-File.

**Comment by jkeenan** [comment]
> @jkeenan I believe that #20753 fixes this.

Unfortunately, it does not.  See https://perl5.test-smoke.org/report/5030348; re-opening ticket.

**Comment by demerphq** [comment]
Hi @jkeenan.  TLDR: what you have reported is not a bug in File::Find and appears to be a race condition in Test::Smoke or how you are using it. I am closing this ticket again.

The smoke report you [linked to](https://perl5.test-smoke.org/file/log_file/5030348) contains this:

```
Can't cd to .. from ./FF_find_t_RW4mMw/for_find/fb/fba: No such file or directory at /usr/home/jkeenan/p5smoke/perl-current/ext/File-Find/../../lib/File/Find.pm line 477.
Failed to chdir to '': No such file or directory at t/taint.t line 104.
END failed--call queue aborted.
# Looks like your test exited with 2 just after 1.
```
This doesn't make any sense and I don't think it is related to this ticket. if you grep t/taint.t for "FF_find_t" you will not find it. You *will* find it in ext/File-Find/t/find.t however. ext/File-Find/t/taint.t uses the temp file prefix "FF_taint_t". 

I do not know what is going on with your smokes, but at this point I am going to blame the smoke code or something else on your box.  It smells to me like a race condition is happening, either inside of Test::Smoke itself or in the way you are running it, or triggered by some other process outside of the perl test code. For instance an outside process could be executing a git clean or something else that cleans or replace your smoke directory with a fresh checkout, and because File::Find is one of the few tests that creates files in the source tree during test it is manifesting by breaking these tests.

Also I do not know what the smoke infra does, I haven't investigated it closely, but it is impossible to debug this without proper test output.  I need to see the full output of `make test` or `make test_harness` I would hope that the smoke infra has a way of storing the actual output of the make test process, so please work with @tux to figure out how to enable it. 

I am going to close this ticket again. Please do not reopen it unless you have actual proof that this bug is related to File::Find and is not an artifact of how you are running the tests or from Test::Smoke itself. At this point I do not trust the output you are providing at all, and what there is strongly suggests a race condition in the smoke infra itself.

Further the error here:  `Failed to chdir to '': No such file or directory at t/taint.t line 104` should not be emitted. From what I can tell from the output you have provided is that the taint.t failed after 1 test. The second test is a duplicate of the first, which checks to see if it can find "taint.t" in the t directory.  It *cannot* fail because of parallelism issues /in the test suite/, it might fail because of parallelism issues *outside* the test suite; for instance the directory being deleted while the tests are running. It is surprising that we don't see any output related to that, but since we can't see the entire test output we don't know if there is more failure output that is simply not being shown. 

If the code died before the second test was run, then it would never set up the directory that the code at line 104 is meant to chdir to, which in turn would mean that we would get the error we see here. We can detect this and I will change the code to do so. However all that will do is prove that this isn't a bug in the test itself.

Anyway, to repeat, nothing in the smoke report you have provided suggests to me that this is a bug in File::Find. It strongly suggests it is a bug or race condition in Test::Smoke or how you are using it. We have had a lot of patches committed recently. If you run your smokes on a cron job and they don't check that there are no smokes already being executed then I could easily imagine this bug happening. 

So I am closing this ticket again.
